packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,153 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/core-utils/internal";
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+	Multiplicity,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export const enum SchemaValidationErrors {
+	NoError,
+	Field_KindNotInSchemaPolicy,
+	Field_IncorrectMultiplicity,
+	Field_NodeTypeNotAllowed,
+	LeafNode_MissingValue,
+	LeafNode_InvalidValue,
+	LeafNode_FieldsNotAllowed,
+	ObjectNode_FieldCountMismatch,
+	ObjectNode_FieldNotInSchema,
+	Node_MissingSchema,
+	UnknownError,
+}
+
+export function isNodeInSchema(
+	node: MapTree,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): SchemaValidationErrors {
+	// Validate the schema declared by the node exists
+	const schema = nodeSchemaCollection.nodeSchema.get(node.type);
+	if (schema === undefined) {
+		return SchemaValidationErrors.Node_MissingSchema;
+	}
+
+	// Validate the node is well formed according to its schema
+
+	if (schema instanceof LeafNodeStoredSchema) {
+		if (node.value === undefined) {
+			return SchemaValidationErrors.LeafNode_MissingValue;
+		}
+		if (node.fields.size !== 0) {
+			return SchemaValidationErrors.LeafNode_FieldsNotAllowed;
+		}
+		if (!allowsValue(schema.leafValue, node.value)) {
+			return SchemaValidationErrors.LeafNode_InvalidValue;
+		}
+	} else if (schema instanceof ObjectNodeStoredSchema) {
+		if (node.fields.size !== schema.objectNodeFields.size) {
+			return SchemaValidationErrors.ObjectNode_FieldCountMismatch;
+		}
+		for (const [fieldKey, field] of node.fields) {
+			const fieldSchema = schema.objectNodeFields.get(fieldKey);
+			if (fieldSchema === undefined) {
+				return SchemaValidationErrors.ObjectNode_FieldNotInSchema;
+			}
+			const fieldInSchemaResult = isFieldInSchema(
+				field,
+				fieldSchema,
+				nodeSchemaCollection,
+				schemaPolicy,
+			);
+			if (fieldInSchemaResult !== SchemaValidationErrors.NoError) {
+				return fieldInSchemaResult;
+			}
```

Comment:
here (and below) you can just return the result, no need to check it, then remove the return SchemaValidationErrors.NoError; case at the bottom.

Reply:
I don't think so... this is inside a for loop over fields, and a single field out of schema means the whole validation can return false, but we could have several valid ones before encountering a bad one, and in that case returning NoError when checking the first one would skip checking any subsequent ones. Or am I missing something?

Reply:
It looks good now, I think I misread this cases and the other ones as well (or they got fixed)